### PR TITLE
Add support for named arguments on special functions

### PR DESCRIPTION
**Pull request by fabpot**

### CODE REVIEW

File: src/ExpressionParser.php [hunk 1]
```diff
@@ -858,4 +853,38 @@ private function checkConstantExpression(Node $node): bool
 
         return true;
     }
+
+    private function parseCalledArguments($arguments, array $names, int $required, string $callType, string $callName, int $line): array
+    {
+        if (count($arguments) > count($names)) {
+            throw new SyntaxError(sprintf('Too many arguments for %s "%s" (it accepts only %d arguments).', $callType, $callName, count($names)), $line, $this->parser->getStream()->getSourceContext());
+        } elseif (count($arguments) < $required) {
```
Comment (stof) [inline src/ExpressionParser.php]: this check is not enough when using named arguments. For instance, `attribute` has 2 required arguments and 1 optional argument. If I pass `variable` and `arguments`, I still don't have less arguments than the number of required arguments, but I don't fill all required arguments.

**Review by stof — CHANGES_REQUESTED** [review]

File: src/ExpressionParser.php [hunk 2]
```diff
@@ -858,4 +853,38 @@ private function checkConstantExpression(Node $node): bool
...
+            throw new SyntaxError(sprintf('Not enough arguments for %s "%s" (it requires at least "%s").', $callType, $callName, implode('", ', array_slice($names, 0, $required))), $line, $this->parser->getStream()->getSourceContext());
```
Comment (smnandre) [inline src/ExpressionParser.php]:
```suggestion
            throw new SyntaxError(sprintf('Not enough arguments for %s "%s" (it requires at least "%s").', $callType, $callName, implode('", "', array_slice($names, 0, $required))), $line, $this->parser->getStream()->getSourceContext());
```

File: src/ExpressionParser.php [hunk 3]
```diff
@@ -858,4 +853,38 @@ private function checkConstantExpression(Node $node): bool
...
+        }
+
+        $args = [];
+        $named = false;
+        foreach ($arguments as $name => $node) {
+            if (!\is_int($name)) {
+                $named = true;
+                $name = $this->normalizeName($name);
+                if (!in_array($name, $names)) {
+                    throw new SyntaxError(\sprintf('Unexpected argument "%s" for %s "%s" (expecting one of "%s").', $name, $callType, $callName, implode('", ', $names)), $line, $this->parser->getStream()->getSourceContext());
```
Comment (smnandre) [inline src/ExpressionParser.php]:
```suggestion
                    throw new SyntaxError(\sprintf('Unexpected argument "%s" for %s "%s" (expecting one of "%s").', $name, $callType, $callName, implode('", "', $names)), $line, $this->parser->getStream()->getSourceContext());
```

File: tests/Fixtures/functions/attribute_with_wrong_args.test [hunk 4]
```diff
@@ -0,0 +1,8 @@
+--TEST--
+"attribute" function
+--TEMPLATE--
+{{ attribute(var=var, template="tpl") }}
+--DATA--
+return ['var' => null]
+--EXCEPTION--
+Twig\Error\SyntaxError: Unexpected argument "var" for function "attribute" (expecting one of "variable", attribute", arguments") in "index.twig" at line 2.
```
Comment (smnandre) [inline tests/Fixtures/functions/attribute_with_wrong_args.test]:
```suggestion
Twig\Error\SyntaxError: Unexpected argument "var" for function "attribute" (expecting one of "variable", "attribute", "arguments") in "index.twig" at line 2.
```

File: src/ExpressionParser.php (hunk 1 above)
Comment (fabpot) [reply src/ExpressionParser.php]: Fixed

**Pull request merged**